### PR TITLE
feat(console sink): accept both logs and metrics

### PR DESCRIPTION
**Pull request by lukesteensen**
### Description

The console sink was documented to accept both logs and metrics, which it should, but we hadn't done the work to support that yet. These changes expand the type system to support a new `Any` type for components that accept logs or metrics, clean up a bunch of stuff around type checking, and add basic support for metric events to the console sink. It'd be best to read them commit-by-commit.

### TODO

- [ ] All commits are signedoff (See CONTRIBUTING.md). Forgot? `make signoff`.
- [ ] CHANGELOG.md has been updated to reflect noteworthy changes.
- [ ] Documentation has been updated to reflect changes (see DOCUMENTING.md).

### CODE REVIEW

File: src/sinks/console.rs [hunk 1]
```diff
@@ -51,22 +56,55 @@ impl crate::topology::config::SinkConfig for ConsoleSinkConfig {
...
+                    .map_err(|e| panic!("Unable to convert json to utf8: {}", e))
+            } else {
+                let s = log
+                    .get(&event::MESSAGE)
+                    .map(|v| v.to_string_lossy())
+                    .unwrap_or_else(|| "".into());
+                Ok(s)
+            }
+        }
+        Event::Metric(metric) => serde_json::to_string(&metric).map_err(|_| ()),
```
Comment (LucioFranco) [inline src/sinks/console.rs]: do we want to attempt to just `fmt::Display` the type if encoding json is not provided?

File: src/topology/config/validation.rs [hunk 2]
```diff
@@ -17,278 +25,247 @@ enum Node {
...
+
+        for (name, config) in config.sinks.iter() {
+            graph.add_sink(name, config.inner.input_type(), config.inputs.clone());
+        }
+
+        graph
+    }
+}
+
+fn paths_rec(nodes: &HashMap<String, Node>, node: &str, mut path: Vec<String>) -> Vec<Vec<String>> {
```
Comment (LucioFranco) [inline src/topology/config/validation.rs]: Does this stand for `paths_recursive`?

File: src/topology/config/validation.rs [hunk 3]
```diff
@@ -17,278 +25,247 @@ enum Node {
...
+            .collect::<HashSet<_>>();
+        let mut edges = self.edges();
+
+        let mut no_incoming = nodes
+            .into_iter()
+            .filter(|n| !edges.iter().any(|(_t, h)| h == n))
+            .collect::<Vec<String>>();
+        while let Some(node) = no_incoming.pop() {
+            let outgoing = edges
+                .clone()
```
Comment (LucioFranco) [inline src/topology/config/validation.rs]: I dont think you need to clone this entire set, you should be able to `iter -> filter -> map(Clone::clone) -> collect` I think.

File: src/topology/config/validation.rs [hunk 4]
```diff
@@ -17,278 +25,247 @@ enum Node {
...
             Err(vec![
                 "Data type mismatch between in (Log) and out (Metric)".into()
             ]),
-            typecheck(&badly_typed)
+            graph.typecheck()
+        );
+    }
+
+    #[test]
+    fn allows_log_or_metric_into_any() {
```
Comment (LucioFranco) [inline src/topology/config/validation.rs]: do we want to add a test around checking any for transforms as well?

File: src/sinks/aws_cloudwatch_logs.rs [hunk 5]
```diff
@@ -107,7 +105,7 @@ pub enum CloudwatchError {
 }
 
 #[typetag::serde(name = "aws_cloudwatch_logs")]
-impl crate::topology::config::SinkConfig for CloudwatchLogsSinkConfig {
+impl SinkConfig for CloudwatchLogsSinkConfig {
```
[the remark this answers is not in the document]
Comment (lukesteensen) [reply src/sinks/aws_cloudwatch_logs.rs]: I knew you'd be excited about this :smile:

File: src/sinks/console.rs (hunk 1 above)
Comment (lukesteensen) [reply src/sinks/console.rs]: We'd have to implement `Display` for `Metric` and I'm not sure that's super worthwhile. My thought process was that we already default to JSON for structured data and metrics are structured data, but also that it's not a huge deal.

File: src/topology/config/validation.rs (hunk 2 above)
Comment (lukesteensen) [reply src/topology/config/validation.rs]: Yep

File: src/topology/config/validation.rs (hunk 3 above)
Comment (lukesteensen) [reply src/topology/config/validation.rs]: Not super worried about performance in this path, but nice catch.

File: src/topology/config/validation.rs (hunk 4 above)
Comment (lukesteensen) [reply src/topology/config/validation.rs]: Yep, will do.

**Pull request merged**